pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +518,152 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
-	for i := start; i < pageEnd; i++ {
-		d := disks[i]
-		nodeList := []string{}
+	numVisited, numNotVisited, numEntries := 0, 0, 0
+	derr := d.listVolumesByResourceGroup(ctx, d.cloud.ResourceGroup, &entries, start, maxEntries, &numVisited, &numNotVisited, &numEntries, nil)
+	if derr != nil {
+		return nil, derr
+	}
+
+	nextTokenString := ""
+	if numNotVisited > 0 {
+		nextTokenString = strconv.Itoa(numVisited)
+	}
+
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesByResourceGroup is a helper function that updates the ListVolumeResponse_Entry slice and returns error if found
+func (d *Driver) listVolumesByResourceGroup(ctx context.Context, resourceGroup string, entries *[]*csi.ListVolumesResponse_Entry, start, maxEntries int, numVisited, numNotVisited, numEntries *int, volSet *map[string]bool) error {
+	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, resourceGroup)
+	if derr != nil {
+		return status.Errorf(codes.Internal, "ListVolumes on rg(%s) failed with error: %v", resourceGroup, derr.Error())
+	}
 
+	if start > *numVisited+len(disks) {
+		return status.Errorf(codes.FailedPrecondition, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
+	}
+	i := start - *numVisited
+	end := len(disks)
+	if maxEntries > 0 && maxEntries < len(disks)-i {
+		end = maxEntries
+	}
+	*numNotVisited = len(disks)
+	for ; i < end; i++ {
+		disk := disks[i]
+		// if given a set of volumes from KubeClient, only continue if the disk can be found in the set
+		if volSet != nil && (*volSet)[*disk.ID] {
+			continue
+		}
 		// HyperVGeneration property is only setup for os disks. Only the non os disks should be included in the list
-		if d.DiskProperties == nil || d.DiskProperties.HyperVGeneration == "" {
-			if d.ManagedBy != nil {
-				attachedNode, err := vmSet.GetNodeNameByProviderID(*d.ManagedBy)
+		if disk.DiskProperties == nil || disk.DiskProperties.HyperVGeneration == "" {
+			nodeList := []string{}
+
+			if disk.ManagedBy != nil {
+				attachedNode, err := d.cloud.VMSet.GetNodeNameByProviderID(*disk.ManagedBy)
 				if err != nil {
-					return nil, err
+					return err
 				}
 				nodeList = append(nodeList, string(attachedNode))
 			}
 
-			entries = append(entries, &csi.ListVolumesResponse_Entry{
+			*entries = append(*entries, &csi.ListVolumesResponse_Entry{
 				Volume: &csi.Volume{
-					VolumeId: *d.ID,
+					VolumeId: *disk.ID,
 				},
 				Status: &csi.ListVolumesResponse_VolumeStatus{
 					PublishedNodeIds: nodeList,
 				},
 			})
+			*numEntries++
 		}
 	}
-
-	nextTokenString := ""
-	if pageEnd < len(disks) {
-		nextTokenString = strconv.Itoa(pageEnd)
-	}
-
-	listVolumesResp := &csi.ListVolumesResponse{
-		Entries:   entries,
-		NextToken: nextTokenString,
-	}
-
-	return listVolumesResp, nil
+	*numVisited, *numNotVisited = *numVisited+i, *numNotVisited-i
```

Comment:
Changing the values of these variables(numVisited, numNotVisited etc) by reference might lead to synchronization issues if multiple go routines are accessing these at the same time. That might make the values of these variables inconsistent across goroutines.
It would be better to pass these variables into this method by value, change(increment/decrement) the values and then send these to the calling method as return values.